coding-the-legislation/25_vectorial_computing.md

```diff
@@ -1,67 +1,183 @@
-This enables fast calculations for a big population, but comes with some constraints.
+OpenFisca calculation are all **vectorial**. That means they operate on arrays rather than single (“scalar”) values.
 
-## Vectors
+The practical benefit is that computations are almost as expensive to do for one entity as they are for several hundred thousands. This is how datasets can be analysed and how reforms can be modelled accurately. However, to support this feature, you will need to apply some constraints on the way you write formulas.
 
-Each computation in OpenFisca returns a **vector**. For instance, for a simulation containing 3 persons whose ages are 45, 42, and 41, executing the following formula:
+
+## Formulas always return vectors
+
+Each computation in OpenFisca must return a vector. For instance, for a simulation containing 3 persons whose ages are 41, 42 and 45, executing the following formula:
 
 ```py
-def formula(person, period):
-    age = person('age', period)
-    print(age)     
+def formula(persons, period, parameters):
+    age = persons('age', period)
+    print(age)
 ```
 
 will print:
+
 ```
-array([45, 42, 17])
+array([41, 42, 45])
 ```
 
-Basic operations such as `+` or `*` behave the same way on vectors than on numbers, you can thus use them in OpenFisca. However, some operations and structures must be adapted.
+This formula code will work the same if there is one Person or three or three million in the modelled situation. Formulas always receive as their first parameter an array of the [entity](./coding-the-legislation/50_entities.md) on which they operate (e.g. *n* Person, Household…) and they should return an array of the same length.
 
-## Scalars
 
-As a first-time developer, it is natural to write such a formula
+Most of the time, formulas will refer to other variables and Numpy will do the appropriate computation without you even noticing:
+
 ```py
-def formula(person):
-    rebate = 610
-    return rebate
+def formula(persons, period, parameters):
+    tax_rebate = parameters(period).tax_rebate  # let's say this is 500
+    eligibility_multiplier = persons('eligibility_multiplier', period)  # and this is [2, 0, 1]: there are three Persons
+    return eligibility_multiplier * tax_rebate  # this is [1000, 0, 500]. We've returned a vector, yay!
 ```
-Unfortunately, this will generate an error similar to
-> The formula ‘rates_rebate@2016’ should return a Numpy array; instead it returned ‘610.0’ of ’<type ‘float’>’.
 
-Indeed, formulas receive an array of entity (*n* _people_ or *m* _households_ in `country_template`) and OpenFisca expects formulas to return an array of the same shape.
+### What happens if you don't return a vector
+
+As programmers, we are used to work with scalars rather than vectors. We thus have a tendency to write straightforward code that returns a scalar rather than a unidimensional vector, and to loop over it:
```

Comment:
This is good; from my own experience where I initially missed that I was dealing with vectors was in my first few attempts to work with the country template and mimic the tests supplied in that - the example tests gave the impression everything was scalar as they were all individual scenarios. This makes sense from a testing perspective but from a learning path perspective it gave a false impression. Perhaps the country template should add to the current tests some tests that run more than one item? This could lead to better testing as I successfully wrote a test that would pass for one item but would fail for many.